### ISSUE TITLE
silabs/gecko/emlib/src/em_gpio.c:111:35: warning: ?: using integer constants in boolean context [-Werror=int-in-bool-context]

### ISSUE THREAD

**Issue by marc-hb** [zephyrproject-rtos/zephyr#16968]
**Describe the bug**
[ccache used to cache pre-processor output as a speed optimization](https://github.com/ccache/ccache/issues/116). But contrary to a popular myth, C compilers don't handle already pre-processed code the same. Feeding the compiler with already pre-processed code surfaces one hidden warning in `silabs/gecko/emlib/src/em_gpio.c`. This used to be a sanitycheck showstopper because sanitycheck adds `-Werror`.
```
modules/hal/silabs/gecko/emlib/src/em_gpio.c: In function 'GPIO_DriveModeSet':
modules/hal/silabs/gecko/emlib/src/em_gpio.c:111:35: warning: ?: using integer constants in boolean context [-Werror=int-in-bool-context]
   EFM_ASSERT(GPIO_PORT_VALID(port) && GPIO_DRIVEMODE_VALID(mode));
           ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

Looking at the pre-processed output, the warning is valid (the name GPIO_PORT_VALID is misleading) and easy to fix with some `(bool)` casts. It looks like the warning gets "optimized away" when not using ccache.

**Workarounds**
- Disable ccache - and everything becomes slower
- `rm -rf ~/.ccache; export CCACHE_CPP2=anything_except_yes` and everything becomes slower but just once.

Second workaround and possible explanation found at http://petereisentraut.blogspot.com/2011/09/ccache-and-clang-part-2.html 

**To Reproduce** on any Unix system with ccache
```
mv ~/.ccache ~/ccache.saved
unset CCACHE_CPP2
export CCACHE_NOCPP2=anything
ccache --print-config | grep cpp # (environment) run_second_cpp = false
cd ${ZEPHYR_BASE}/samples/synchronization
rm -rf build/
cmake -B build -S . -DBOARD=efm32hg_slstk3400a -DEXTRA_CFLAGS=-Werror
make -C build VERBOSE=1
```

**Expected behavior**
Compiles.

**Impact**
sanitycheck showstopper

**Environment (please complete the following information):**
 - OS: Ubuntu 16.04, ccache version 3.2.4
 - Toolchain  Zephyr SDK 0.10.0
- https://github.com/zephyrproject-rtos/hal_silabs 9151e614c239

**Additional context**
I verified that ccache doesn't change the .obj file, it only adds the warning.

**Comment by marc-hb** [comment]
Following a suggestion from @joshtriplett I put ccache aside and separated the pre-processing and actual compilation steps:
```
$ arm-zephyr-eabi-gcc <long arg list copied from VERBOSE=1> -c em_gpio.c -E -o em_gpio.i
$ arm-zephyr-eabi-gcc <long arg list copied from VERBOSE=1> -c em_gpio.i 
```
This does produce the same warning on the second, actual compilation step. This confirms that the Zephyr SDK behaves somewhat differently when performing both passes at once and that the http://petereisentraut.blogspot.com/2011/09/ccache-and-clang-part-2.html article is not clang specific.

Whether it gets "optimized away" or not and when, the warning seems valid and the simplest seems to just fix it.

**Comment by marc-hb** [comment]
I found why I hit this issue only on Ubuntu 16.04 and why sanitycheck doesn't fail on Ubuntu 18.04 or Fedora: newer distributions changed the default ccache setting:
```
ubuntu16$ ccache --print-config | grep cpp
(default) run_second_cpp = false
```
Description updated accordingly.

**Comment by chrta** [comment]
Info: The current gecko sdk (emlib) 5.8.0 still contains this issue.

**Comment by chrta** [comment]
Since i neither use ubuntu 16.04 nor ccache, i had to use seperate preprocessing to reproduce your problem.
Neither bool-casts, nor adding a `!!` fixed the issue for me, only this:
```diff
-#define GPIO_PORT_VALID(port)          (_GPIO_PORT_MASK(port) )
+#define GPIO_PORT_VALID(port)          (_GPIO_PORT_MASK(port) != 0)
```

@marc-hb Could you please verify that this change fixes the issue for you?

**Comment by marc-hb** [comment]
I just tested it and I confirm the following change fixes the compilation issue reported (and the mismatch between the macro name and its previous code)
```diff
--- a/gecko/emlib/inc/em_gpio.h
+++ b/gecko/emlib/inc/em_gpio.h
@@ -406,7 +406,7 @@ extern "C" {
     : 0)
 
 /** Validation of port and pin. */
-#define GPIO_PORT_VALID(port)          (_GPIO_PORT_MASK(port) )
+#define GPIO_PORT_VALID(port)          (_GPIO_PORT_MASK(port) != 0)
 #define GPIO_PORT_PIN_VALID(port, pin) (((_GPIO_PORT_MASK(port)) >> (pin)) & 0x1)
 
 #if defined(_GPIO_EXTIPINSELL_MASK)
```
Disclaimer: compilation is all I tested. I don't see how this could affect anything else though.

> Since i neither use ubuntu 16.04 nor ccache, i had to use seperate preprocessing to reproduce your problem.

The reproduction script I put in the description works on any Linux distribution. What operating system do you use? On pretty much any Linux distribution using ccache is as simple as `apt-get install ccache` or `dnf install ccache` and done, no configuration required. To stop using just uninstall ccache.

**Comment by galak** [comment]
Is this still an issue?  We don't see any failures w/sanitycheck building gpio support on Silabs.

**Comment by marc-hb** [comment]
tl:dr: if you don't care about the programming error and only care about sanitycheck showing green lights then feel free to close as wontfix. Otherwise apply the trivial one-line fix above.

The warning is still valid and the code is still wrong.

However in 2016 ccache disabling its pre-processing optimization in order to align with compilers and show fewer warnings. See https://github.com/ccache/ccache/issues/93

So no one using a recent ccache version will ever see this (valid!) warning.

The `GPIO_PORT_VALID()` macros still returns an integer instead of a boolean:
 https://github.com/zephyrproject-rtos/hal_silabs/blob/78da967feeac0d5/gecko/emlib/inc/em_gpio.h#L433

`em_gpio.c` still uses it as a boolean:
 https://github.com/zephyrproject-rtos/hal_silabs/blob/78da967feeac0d512/gecko/emlib/src/em_gpio.c#L105

PS: I clarified the very short summary at the top.

**Comment by chrta** [comment]
This bug was fixed by merging #27042
